### PR TITLE
Handle Out-of-scope separately in Embedding Classifier Training

### PR DESCRIPTION
**Pull request by georgwiese**
**Proposed changes**:
- So far, out-of-scope examples are treated like a regular intent. I added an option to `EmbeddingIntentClassifier` so that one can specify the name of the out-of-scope "intent".
- If it is enabled, the loss function is changed such that out-of-scope examples are not pushed towards one vector, but still are pushed away from all other intent vectors.
- At prediction time, the classifier predicts out-of-scope if the similarity of the highest-ranking intent is below a threshold.

In my understanding, this should perform better in cases where out-of-scope is very diverse (which it typically is). Also, I think this would more likely classify messages as out-of-scope that are completely unlike any training example.

I don't consider the code final, it still contains some hacks. However, it shouldn't break anything if the option is disabled (it is by default). Also, it should be good enough to test this on a few datasets to decide if we want to merge it.

Let me know if you need something from my side.

**Status (please check what you already did)**:
- [ ] made PR ready for code review
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog

### CODE REVIEW

File: rasa_nlu/classifiers/embedding_intent_classifier.py [hunk 1]
```diff
@@ -347,8 +354,29 @@ def _tf_sim(self, a, b):
...
+            For each element of the batch, the similarity of the correct
+            intent to the num_neg negative intents.
+        - similarity_mask:
+            Tensor of shape [batch_size]
+            Indicates whether the example's similarity to its intent should
+            contribute to the loss.
+        """
+
+        # Set masked similarities to 1
+        sim = tf.concat([
```
Comment (tmbo) [inline rasa_nlu/classifiers/embedding_intent_classifier.py]: @georgwiese can you put an assert in here that `sim` is the same before and after this statement if `similarity_mask` is set to all ones (which reflects deactivating the feature right?)

**Comment by twhughes** [comment]
If possible, I'd also recommend finding a way to make the code fail more gracefully if `out_of_scope_intent` is not present in the test or train set.

Right now one can specify an `out_of_scope_intent` that is not present in the training set and nlu will train fine and not give a warning.  However, it throws an error if `out_of_scope_intent` isn't present in the test set.

This isn't a huge issue but it would be nice if it gave a warning or error during training.

File: rasa_nlu/classifiers/embedding_intent_classifier.py (hunk 1 above)
Comment (georgwiese) [reply rasa_nlu/classifiers/embedding_intent_classifier.py]: I refactored the code so that `similarity_mask` is optional, and `sim` remains unchanged if it is not provided. Also, I now log a warning if a similarity mask is applied. See 44613833dfc7b1687e664035bed754cc773b69e1.

**Comment by georgwiese** [comment]
I refactored the code a bit to make sure the TF graph is exactly the same as on `master` if the feature is deactivated (as per default). Also, I merged in the current master.

A few notes:
- I wouldn't consider this ready to merge, but it should be enough to test if this feature makes any difference on any of your datasets. I guess if not, I shouldn't bother polishing the PR.
- The `out_of_scope_soft_threshold` is new hyper parameter that potentially needs tuning. On our dataset, the default (`0.5`) was a good choice, I'm not sure how much this potentially varies from dataset to dataset.

**Comment by Ghostvv** [comment]
Shouldn't we instead of `out_of_scope_soft_threshold` during prediction check if max similarity is below `-mu_neg+eps`?

**Comment by georgwiese** [comment]
So then the current code with `out_of_scope_soft_threshold=0.5` and `mu_neg=-0.4` (defaults) would be equivalent to that calculation with `eps=0.1`, right?

Seems reasonable to make the threshold relative to `mu_neg`. Maybe another reasonable default would be the average of `-mu_neg` and `mu_pos`. Not sure what's the best choice.

**Comment by twhughes** [comment]
Thanks Georg, I will re-test either today or tomorrow.

**Comment by twhughes** [comment]
@georgwiese , I tried the new commit on each of our oos datasets and didn't observe any change in OOS detection, unfortunately.

I am going to do some experiments with the intent vectors themselves to see if we can learn anything about whether the new loss function is doing what we expect.  I'll keep you updated on that.

**Comment by georgwiese** [comment]
Sure, but I only added two lines:

```yaml
language: "de"

pipeline:
- name: "intent_featurizer_count_vectors"
- name: "intent_classifier_tensorflow_embedding"
  embed_dim: 20
  num_neg: 20
  epochs: 200
  out_of_scope_training_flag: true
  out_of_scope_intent: request_outofscope
```

Adding the last two lines and mounting my fork are the only two changes I've made to our repo.

File: rasa_nlu/classifiers/embedding_intent_classifier.py [hunk 2]
```diff
@@ -347,8 +354,35 @@ def _tf_sim(self, a, b):
...
+
+        # Set masked similarities to 1
+        if similarity_mask is not None:
+            # For debugging:
+            # similarity_mask = tf.Print(similarity_mask,
+            #     [tf.reduce_sum(tf.cast(similarity_mask, tf.int32))],
+            #     message="Number of masked similarities")
+            logging.warn("Masking some similarities")
+            sim = tf.concat([
+                tf.where(similarity_mask, sim[:, 0:1], tf.ones((tf.shape(sim)[0], 1), tf.float32)),
```
[the remark this answers is not in the document]
Comment (georgwiese) [reply rasa_nlu/classifiers/embedding_intent_classifier.py]: The loss should always be 0 for a similarity of 1, right?
[Line 389](https://github.com/RasaHQ/rasa_nlu/pull/1229/files/158f8cafd259402646eebdbbde7f534932560bba#diff-db349250de8b693341954da8642144f5R389) computes `max(0, mu_pos - sim)`, which is 0 for all similarities above `mu_pos`, including 1.

[the remark this answers is not in the document]
Comment (georgwiese) [reply rasa_nlu/classifiers/embedding_intent_classifier.py]: Yes, but why would `mu_pos` be larger than 1 if the cosine similarity can't be bigger than 1? I think a perfect cosine similarity of 1 should always lead to a loss of 0, right?

[the remark this answers is not in the document]
Comment (Ghostvv) [reply rasa_nlu/classifiers/embedding_intent_classifier.py]: you're right, but if the user wanted to use inner product as similarity

**Comment by tmbo** [comment]
@Ghostvv what is the status on this one?

**Comment by Ghostvv** [comment]
I think our experiments didn't show any improvement
